In [1]:
!pip install numpy==1.23.1
!pip install pandas==1.4.3
!pip install matplotlib==3.3.2
!pip install seaborn==0.11.0
!pip install scikit_learn==1.0.2
!pip install scipy==1.9.0

  Using cached numpy-1.23.1-cp39-cp39-win_amd64.whl (14.7 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.22.4
    Uninstalling numpy-1.22.4:
      Successfully uninstalled numpy-1.22.4


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
daal4py 2021.3.0 requires daal==2021.2.3, which is not installed.
numba 0.54.1 requires numpy<1.21,>=1.17, but you have numpy 1.23.1 which is incompatible.
scipy 1.7.1 requires numpy<1.23.0,>=1.16.5, but you have numpy 1.23.1 which is incompatible.


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


  Using cached matplotlib-3.3.2.tar.gz (37.9 MB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Running setup.py clean for matplotlib
Failed to build matplotlib


  error: subprocess-exited-with-error
  
  python setup.py bdist_wheel did not run successfully.
  exit code: 1
  
  [571 lines of output]
  
  Edit setup.cfg to change the build options; suppress output with --quiet.
  
  BUILDING MATPLOTLIB
    matplotlib: yes [3.3.2]
        python: yes [3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit
                    (AMD64)]]
      platform: yes [win32]
   sample_data: yes [installing]
         tests: no  [skipping due to configuration]
        macosx: no  [Mac OS-X only]
  
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.9
  copying lib\pylab.py -> build\lib.win-amd64-3.9
  creating build\lib.win-amd64-3.9\matplotlib
  copying lib\matplotlib\afm.py -> build\lib.win-amd64-3.9\matplotlib
  copying lib\matplotlib\animation.py -> build\lib.win-amd64-3.9\matplotlib
  copying lib\matplotlib\artist.py -> build\lib.win-amd64-3.9\matplotlib
  copying lib\matplotlib\axis.py -> build\li

  Using cached numpy-1.22.4-cp39-cp39-win_amd64.whl (14.7 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.23.1
    Uninstalling numpy-1.23.1:
      Successfully uninstalled numpy-1.23.1


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
daal4py 2021.3.0 requires daal==2021.2.3, which is not installed.
numba 0.54.1 requires numpy<1.21,>=1.17, but you have numpy 1.22.4 which is incompatible.


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


     -------------------------------------- 38.6/38.6 MB 747.7 kB/s eta 0:00:00
  Attempting uninstall: scipy
    Found existing installation: scipy 1.7.1
    Uninstalling scipy-1.7.1:
      Successfully uninstalled scipy-1.7.1


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


In [2]:
import numpy as np
import pandas as pd
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)

import matplotlib.pyplot as plt
import seaborn as sns

import os
import time 
import re

import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn import metrics

from scipy import stats

import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('dataset/news_article_ready.csv')
print(df.shape)
df.head()

(823, 42)


unique_tokens_rate  num_hrefs  num_imgs  num_videos  average_token_length  \
0            0.640816        5.0       3.0         1.0              4.579592   
1            0.549618        4.0       1.0         0.0              4.636704   
2            0.577236        8.0       1.0         0.0              4.627451   
3            0.525896        4.0       1.0         0.0              5.083665   
4            0.588889        6.0       9.0         0.0              4.336022   

   num_keywords  min_avg_key  max_avg_key  avg_avg_key  href_avg_shares  \
0           5.0     1459.500     4817.692     3370.415         1833.333   
1           7.0        0.000     3581.094     2556.367         8500.000   
2           7.0     2270.889     3852.800     3018.167         2076.333   
3           9.0        0.000     3581.094     1871.403         2200.000   
4           7.0     3581.094     8453.451     5155.323          805.000   

   global_subjectivity  global_sentiment_polarity  global_rate_positive_words  \
0             0.403762                   0.038762                    0.036735   
1             0.410707                   0.180934                    0.044944   
2             0.397545                   0.044864                    0.035294   
3             0.491578                   0.157620                    0.047809   
4             0.628535                   0.078497                    0.053763   

   global_rate_negative_words  title_subjectivity  title_sentiment_polarity  \
0                    0.024490                0.25                      0.25   
1                    0.003745                0.00                      0.00   
2                    0.019608                0.00                      0.00   
3                    0.011952                0.00                      0.00   
4                    0.034946                0.00                      0.00   

   shares  title_Word_Count  text_Word_Count  is_weekday  not_weekday  \
0  1600.0               7.0            176.0         1.0          0.0   
1  1500.0              11.0             86.0         1.0          0.0   
2  1400.0              14.0            439.0         1.0          0.0   
3  1300.0               6.0            241.0         1.0          0.0   
4  3000.0               9.0            258.0         1.0          0.0   

   average_shares_by_channel_date  text_readability_score  \
0                     1600.000000                     8.3   
1                     2402.631579                     6.6   
2                     2402.631579                     9.6   
3                     2402.631579                     8.0   
4                     1815.000000                    11.6   

   title_readability_score  noun_count  verb_count  adjective_count  \
0                      2.1        52.0        21.0             22.0   
1                      3.7        24.0        11.0             14.0   
2                      4.0       127.0        48.0             62.0   
3                      8.0        71.0        29.0             28.0   
4                      4.4        69.0        34.0             35.0   

   april opens  last year  new window  new york  nggallery id  opens new  \
0          0.0        0.0         0.0       0.0           0.0        0.0   
1          0.0        0.0         0.0       0.0           1.0        0.0   
2          5.0        0.0        17.0       0.0           0.0       15.0   
3          0.0        0.0         2.0       0.0           0.0        2.0   
4          0.0        0.0         1.0       0.0           0.0        1.0   

   pic twitter  seealso slug  social media  twitter com  \
0          0.0           1.0           0.0          0.0   
1          0.0           0.0           0.0          0.0   
2          5.0           0.0           0.0          5.0   
3          0.0           1.0           0.0          0.0   
4          0.0           1.0           0.0          0.0   

   data_channel_business  data_channel_entertainment  data_chann

In [20]:
X = df.drop('shares',axis=1)
y = df[['shares']]

print(X.shape)
print(y.shape)

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=163)
print((X_train.shape, X_test.shape), (y_train.shape,y_test.shape))

(823, 41)
(823, 1)
((576, 41), (247, 41)) ((576, 1), (247, 1))


# Model Training

## Linear Regression

In [6]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [21]:
# Model training

# Create a Linear Regression model
linear_reg = LinearRegression()

# Fit the model to the training data
linear_reg.fit(X_train, y_train)

y_train_pred = linear_reg.predict(X_train)

# Evaluate the trained model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
print("Linear Training RMSE:", train_rmse)

train_r2 = r2_score(y_train, y_train_pred)
print("Linear Training R^2 Score:", train_r2)

Linear Training RMSE: 2564.9117200471883
Linear Training R^2 Score: 0.19544588443326338


In [22]:
# Make predictions on the testing data
y_test_pred = linear_reg.predict(X_test)

# Evaluate the model
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
print("Linear Testing RMSE:", test_rmse)

test_r2 = r2_score(y_test, y_test_pred)
print("Linear Testing R^2 Score:", test_r2)

Linear Testing RMSE: 2486.764220656687
Linear Testing R^2 Score: 0.14701627771388792


#### Note: RMSE is very high, means this model's predictions are far from the actual values. R^2 also far from 1 and 0.147 is relatively low, indicating that there is a lot of unexplained variance in the data.

### Polynomial Regression

In [17]:
from sklearn.preprocessing import PolynomialFeatures

# Create polynomial features
degree = 1 
poly = PolynomialFeatures(degree=degree)
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.transform(x_test)

# Create a Polynomial Regression model
poly_reg = LinearRegression()

# Fit the model to the training data with polynomial features
poly_reg.fit(x_train_poly, y_train)

# Make predictions on the training and testing data
y_train_pred = poly_reg.predict(x_train_poly)
y_test_pred = poly_reg.predict(x_test_poly)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Polynomial Training RMSE:", train_rmse)
print("Polynomial Testing RMSE:", test_rmse)
print("Polynomial Training R^2 Score:", train_r2)
print("Polynomial Testing R^2 Score:", test_r2)


Polynomial Training RMSE: 2772.68110595552
Polynomial Testing RMSE: 2685.2704980016565
Polynomial Training R^2 Score: 0.059821614378249954
Polynomial Testing R^2 Score: 0.005401948458466643


In [18]:
# Create polynomial features
degree = 2 
poly = PolynomialFeatures(degree=degree)
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.transform(x_test)

# Create a Polynomial Regression model
poly_reg = LinearRegression()

# Fit the model to the training data with polynomial features
poly_reg.fit(x_train_poly, y_train)

# Make predictions on the training and testing data
y_train_pred = poly_reg.predict(x_train_poly)
y_test_pred = poly_reg.predict(x_test_poly)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Polynomial Training RMSE:", train_rmse)
print("Polynomial Testing RMSE:", test_rmse)
print("Polynomial Training R^2 Score:", train_r2)
print("Polynomial Testing R^2 Score:", test_r2)


Polynomial Training RMSE: 0.0003442472410712312
Polynomial Testing RMSE: 70383.51060364797
Polynomial Training R^2 Score: 0.9999999999999855
Polynomial Testing R^2 Score: -682.3030715282749


In [19]:
# Create polynomial features
degree = 3 
poly = PolynomialFeatures(degree=degree)
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.transform(x_test)

# Create a Polynomial Regression model
poly_reg = LinearRegression()

# Fit the model to the training data with polynomial features
poly_reg.fit(x_train_poly, y_train)

# Make predictions on the training and testing data
y_train_pred = poly_reg.predict(x_train_poly)
y_test_pred = poly_reg.predict(x_test_poly)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Polynomial Training RMSE:", train_rmse)
print("Polynomial Testing RMSE:", test_rmse)
print("Polynomial Training R^2 Score:", train_r2)
print("Polynomial Testing R^2 Score:", test_r2)


Polynomial Training RMSE: 8.505064539318223e-08
Polynomial Testing RMSE: 637720.8460632055
Polynomial Training R^2 Score: 1.0
Polynomial Testing R^2 Score: -56095.11056380086


#### NOTE: The scores indicate that Polynomial Regression model has overfit the training data. While it fits the training data extremely well (very low RMSE and very high R^2 on training data) when degree is high, but it fails to generalize to new data (high RMSE and negative R^2 on testing data). If we reduce the degree it underfits.

## Ridge Regression

In [29]:
from sklearn.linear_model import Ridge

# Create a Ridge Regression model
alpha = 0.10
ridge_reg = Ridge(alpha=alpha)

# Fit the model to the training data
ridge_reg.fit(X_train, y_train)

# Make predictions on the training and testing data
y_train_pred = ridge_reg.predict(X_train)
y_test_pred = ridge_reg.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Ridge Training RMSE:", train_rmse)
print("Ridge Testing RMSE:", test_rmse)
print("Ridge Training R^2 Score:", train_r2)
print("Ridge Testing R^2 Score:", test_r2)


Ridge Training RMSE: 2569.5885591669858
Ridge Testing RMSE: 2500.4965498059073
Ridge Training R^2 Score: 0.19250917465533335
Ridge Testing R^2 Score: 0.13756962815816443


#### NOTE: Even if we hypertune the alpha value, the diffrence isn't very noticable. The model isn't working good.

## ElasticNet Regression

In [38]:
from sklearn.linear_model import ElasticNet

# Create an ElasticNet Regression model
alpha = 0.00001 
l1_ratio = .0005 
elastic_net = ElasticNet(alpha=alpha, l1_ratio=l1_ratio)

# Fit the model to the training data
elastic_net.fit(X_train, y_train)

# Make predictions on the training and testing data
y_train_pred = elastic_net.predict(X_train)
y_test_pred = elastic_net.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("ElasticNet Training RMSE:", train_rmse)
print("ElasticNet Testing RMSE:", test_rmse)
print("ElasticNet Training R^2 Score:", train_r2)
print("ElasticNet Testing R^2 Score:", test_r2)

ElasticNet Training RMSE: 2565.1969428208254
ElasticNet Testing RMSE: 2489.0613283669345
ElasticNet Training R^2 Score: 0.19526693876918877
ElasticNet Testing R^2 Score: 0.14543969039721705


#### NOTE: ElasticNet Regression also not fitting good.

## Support Vector Regression (SVR)

In [44]:
from sklearn.svm import SVR

# Create an SVR model
svr = SVR(C=0.005, epsilon=0.001)

# Fit the model to the training data
svr.fit(X_train, y_train) 

# Make predictions on the training and testing data
y_train_pred = svr.predict(X_train)
y_test_pred = svr.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("SVR Training RMSE:", train_rmse)
print("SVR Testing RMSE:", test_rmse)
print("SVR Training R^2 Score:", train_r2)
print("SVR Testing R^2 Score:", test_r2)


SVR Training RMSE: 3072.176500014465
SVR Testing RMSE: 2957.737129833221
SVR Training R^2 Score: -0.15425759869932132
SVR Testing R^2 Score: -0.20667598537650345


#### NOTE: Not good....

## Decision Tree

In [55]:
from sklearn.tree import DecisionTreeRegressor

# Create a Decision Tree model
max_depth = 5
decision_tree = DecisionTreeRegressor(max_depth=max_depth, random_state=42)

# Fit the model to the training data
decision_tree.fit(X_train, y_train)

# Make predictions on the training and testing data
y_train_pred = decision_tree.predict(X_train)
y_test_pred = decision_tree.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Decision Tree Training RMSE:", train_rmse)
print("Decision Tree Testing RMSE:", test_rmse)
print("Decision Tree Training R^2 Score:", train_r2)
print("Decision Tree Testing R^2 Score:", test_r2)

Decision Tree Training RMSE: 1840.9777601974383
Decision Tree Testing RMSE: 3423.2005645780755
Decision Tree Training R^2 Score: 0.5855162573327869
Decision Tree Testing R^2 Score: -0.6163529991491057


#### NOTE: Still not good...

## Gradient Boosting

In [56]:
from sklearn.ensemble import GradientBoostingRegressor

# Create a Gradient Boosting model
n_estimators = 100 
learning_rate = 0.1  
max_depth = 4
gradient_boosting = GradientBoostingRegressor(
    n_estimators=n_estimators,
    learning_rate=learning_rate,
    max_depth=max_depth,
    random_state=42
)

# Fit the model to the training data
gradient_boosting.fit(X_train, y_train)

# Make predictions on the training and testing data
y_train_pred = gradient_boosting.predict(X_train)
y_test_pred = gradient_boosting.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Gradient Boosting Training RMSE:", train_rmse)
print("Gradient Boosting Testing RMSE:", test_rmse)
print("Gradient Boosting Training R^2 Score:", train_r2)
print("Gradient Boosting Testing R^2 Score:", test_r2)


Gradient Boosting Training RMSE: 815.8369995266457
Gradient Boosting Testing RMSE: 2733.932666460235
Gradient Boosting Training R^2 Score: 0.9186013482535882
Gradient Boosting Testing R^2 Score: -0.030972667262670894


## Random Forests

#### NOTE: Prone to overfit. Working well on training data, bad on testing. Even if I hypertune, still overfits.

In [58]:
from sklearn.ensemble import RandomForestRegressor

# Create a Random Forest model
n_estimators = 100  
max_depth = None  # Maximum depth (None for full depth)
random_forest = RandomForestRegressor(
    n_estimators=n_estimators,
    max_depth=max_depth,
    random_state=42
)

# Fit the model to the training data
random_forest.fit(X_train, y_train)  # ravel() is used to convert y_train to a 1D array

# Make predictions on the training and testing data
y_train_pred = random_forest.predict(X_train)
y_test_pred = random_forest.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("Random Forest Training RMSE:", train_rmse)
print("Random Forest Testing RMSE:", test_rmse)
print("Random Forest Training R^2 Score:", train_r2)
print("Random Forest Testing R^2 Score:", test_r2)


Random Forest Training RMSE: 1055.354653752495
Random Forest Testing RMSE: 2618.9619373507326
Random Forest Training R^2 Score: 0.863790538536335
Random Forest Testing R^2 Score: 0.05391556721223356


#### NOTE: Did well to some extent compared to other models.

## K-Nearest Neighbors (KNN)

In [62]:
from sklearn.neighbors import KNeighborsRegressor

# Create a K-Nearest Neighbors (KNN) model
n_neighbors = 2
knn = KNeighborsRegressor(n_neighbors=n_neighbors)

# Fit the model to the training data
knn.fit(X_train, y_train)

# Make predictions on the training and testing data
y_train_pred = knn.predict(X_train)
y_test_pred = knn.predict(X_test)

# Evaluate the model
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

# Print the evaluation scores
print("KNN Training RMSE:", train_rmse)
print("KNN Testing RMSE:", test_rmse)
print("KNN Training R^2 Score:", train_r2)
print("KNN Testing R^2 Score:", test_r2)


KNN Training RMSE: 1747.9510388455933
KNN Testing RMSE: 3136.5545655241076
KNN Training R^2 Score: 0.6263465952313161
KNN Testing R^2 Score: -0.3569917821208246


#### NOTE: Not so good..

## In conclusion among all the models Random Forests did best, but still tot that good. From all of this hassel i can come to this conclusion that maybe my pre-processing wasn't well done, that's why i'm not getting a good result.